From 895b9c2629e2267a1b89ddfa8eedc9349a4fa397 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Tue, 5 Jan 2016 13:21:46 +0100
Subject: [PATCH] video: mxc_hdmi: Improve initial mode setup

Use mxc_fb_find_nearest_mode() for parsing the mode specified on kernel
command line. This makes sure that the interlaced flag is honoured. Use
standard VGA in case the command line argument cannot be parsed.

Also adapt to the fact, that the fb_var_screeninfo returned fb_find_mode()
has it's refresh rate doubled for interlaced modes.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/mxc/mxc_hdmi.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/video/mxc/mxc_hdmi.c b/drivers/video/mxc/mxc_hdmi.c
index 6baf1b6..620168a 100644
--- a/drivers/video/mxc/mxc_hdmi.c
+++ b/drivers/video/mxc/mxc_hdmi.c
@@ -2819,13 +2819,15 @@ static int mxc_hdmi_disp_init(struct mxc_dispdrv_handle *disp,
 
 	/* Set the default mode when disp init. */
 	fb_find_mode(&hdmi->fbi->var, hdmi->fbi,
-		     hdmi->dft_mode_str, NULL, 0, NULL,
-		     hdmi->default_bpp);
+		     hdmi->dft_mode_str, NULL, 0,
+		     &vga_mode, hdmi->default_bpp);
 	fb_var_to_videomode(&m, &hdmi->fbi->var);
+	if (m.vmode & FB_VMODE_INTERLACED)
+		m.refresh = (m.refresh / 2) - 2;
 	hdmi->dft_mode_set = false;
 
 	/* Find a nearest mode in default modelist */
-	mode = fb_find_nearest_mode(&m, &hdmi->fbi->modelist);
+	mode = mxc_fb_find_nearest_mode(&m, &hdmi->fbi->modelist);
 	if (!mode) {
 		pr_err("%s: could not find mode in modelist\n", __func__);
 		return -1;
-- 
1.9.1

